### PR TITLE
Login System

### PR DESCRIPTION
**Pull request by shiara02**

### CODE REVIEW

File: html/login/controllers/authentication.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<?php
+require_once "../models/user.php";
+
+class AuthenticationController {
```
Comment (hipe01mark) [inline html/login/controllers/authentication.php]: Please follow PSR 4.

https://www.php-fig.org/psr/psr-4/examples/

Comment (hipe01mark) [reply html/login/controllers/authentication.php]: check also other lines

File: html/login/models/user.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<?php
+require_once "../config/database.php";
+
+class UserModel {
```
Comment (hipe01mark) [inline html/login/models/user.php]: Follow PSR 4 
https://www.php-fig.org/psr/psr-4/examples/

File: html/login/.env [hunk 3]
```diff
@@ -0,0 +1,4 @@
+DB_SERVERNAME=mysql-server
```
Comment (bpocrafael) [inline html/login/.env]: why is this included in this repo?

.env should never be on the repo because you are going to expose private credentials

File: html/index.php [hunk 4]
```diff
@@ -1,14 +1,19 @@
...
+    <a href="HIPE-EXERCISE/PHP-3/fibonacci.php">PHP-3: Fibonacci Series</a><br>
+    <a href="HIPE-EXERCISE/PHP-4/simple_login.php">PHP-4: Simple Login System</a><br>
+    <a href="HIPE-EXERCISE/PHP-5/form_validation.php">PHP-5: Form Validation</a><br>
+    <a href="HIPE-EXERCISE/PHP-6/array_manipulation.php">PHP-6: Array Manipulation</a><br>
+    <a href="HIPE-EXERCISE/PHP-7/contact_form.php">PHP-7: Simple Contact Form</a><br>
+    <a href="HIPE-EXERCISE/PHP-8/date_time.php">PHP-8: Working with Date and Time</a><br>
+    <a href="HIPE-EXERCISE/PHP-9/string_manipulation.php">PHP-9: String Manipulation</a><br>
+    <a href="HIPE-EXERCISE/PHP-10/basic_file_operations.php">PHP-10: Basic File Operations</a><br>
+</body>
+</html>
```
Comment (bpocrafael) [inline html/index.php]: add new line

File: html/login/config/database.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        // Use environment variables for database connection
+        $servername = $envVariables['DB_SERVERNAME'];
+        $username = $envVariables['DB_USERNAME'];
+        $password = $envVariables['DB_PASSWORD'];
+        $dbname = $envVariables['DB_DBNAME'];
+
+        // Initializes the db connection
+        $this->conn = new mysqli($servername, $username, $password, $dbname);
+
+        // Check if the connection was successful
```
Comment (bpocrafael) [inline html/login/config/database.php]: this is understandable why do you keep commeting.

File: html/login/config/database.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        } else {
+            die('.env file not found.');
+        }
+        // Use environment variables for database connection
+        $servername = $envVariables['DB_SERVERNAME'];
+        $username = $envVariables['DB_USERNAME'];
+        $password = $envVariables['DB_PASSWORD'];
+        $dbname = $envVariables['DB_DBNAME'];
+
+        // Initializes the db connection
```
Comment (bpocrafael) [inline html/login/config/database.php]: nit, remove comments

File: html/login/controllers/authentication.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                if (!isset($_POST['username']) || !isset($_POST['password'])) {
+                    throw new Exception("Username and password are required.");
+                }
+
+                $username = $_POST['username'];
+                $password = $_POST['password'];
+
+                if (empty($username)) {
+                    throw new Exception("Username is required");
+                } elseif (empty($password)) {
```
Comment (bpocrafael) [inline html/login/controllers/authentication.php]: can you separate the condition of this.

File: html/login/controllers/authentication.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+class AuthenticationController {
+    public function authenticate($username, $password) {
+        $userModel = new UserModel();
+        $user = $userModel->authenticateUser($username, $password);
+
+        if ($user) {
+            header("Location: ../views/dashboard.php");
+            exit();
+        } else {
+            // Redirect to the login page with an error message
```
Comment (bpocrafael) [inline html/login/controllers/authentication.php]: remove comment

File: html/login/views/dashboard.php [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+<html>
+<head>
+    <title>Dashboard</title>
+    <link rel="stylesheet" type="text/css" href="../public/style.css">
+</head>
+<body>
+    <h1>Welcome to the Dashboard</h1><br>
+    <a href="../views/login.php">Logout</a>
+</body>
+</html>
```
Comment (bpocrafael) [inline html/login/views/dashboard.php]: add new line

File: html/login/views/login.php [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+<html>
+    <head>
+        <title>LOGIN</title>
+        <link rel="stylesheet" type="text/css" href="../public/style.css">
+    </head>
+    <body>
+        <form action="../controllers/authentication.php" method="post">
+            <h2>LOGIN</h2>
+            <?php if (isset($_GET['error'])) { ?>
+                <p class="error"><?php echo $_GET['error']; ?></p>
```
Comment (bpocrafael) [inline html/login/views/login.php]:
```
<p class="error"><?= $_GET['error']; ?></p>
```

nit, for short syntax and much btter when printing in a html or a template.

always use the shorthand for echoing.

**Review by bpocrafael — CHANGES_REQUESTED** [review]

File: html/login/.env (hunk 3 above)
Comment (bpocrafael) [reply html/login/.env]: you could use .gitignore file to prevent this from adding the .env file

File: html/login/controllers/authentication.php (hunk 7 above)
Comment (shiara02) [reply html/login/controllers/authentication.php]: Yes sir, i already resolved and pushed it. Thank you.

File: html/login/controllers/authentication.php (hunk 8 above)
Comment (shiara02) [reply html/login/controllers/authentication.php]: Yes sir, i already resolved and pushed it. Thank you.

File: html/login/views/dashboard.php (hunk 9 above)
Comment (shiara02) [reply html/login/views/dashboard.php]: Yes sir, i already resolved and pushed it. Thank you.

File: html/login/views/login.php (hunk 10 above)
Comment (shiara02) [reply html/login/views/login.php]: Yes sir, i already resolved and pushed it. Thank you.

File: html/login/controllers/authentication.php (hunk 1 above)
Comment (shiara02) [reply html/login/controllers/authentication.php:6]: Yes sir, i already resolved and pushed it. Thank you.

**Review by bpocrafael — APPROVED** [review]
LGTM 🚢  just added new comment nitpick

File: html/login/models/user.php (hunk 2 above)
Comment (shiara02) [reply html/login/models/user.php:5]: Yes sir, i already resolved and pushed it. Thank you.

File: html/login/controllers/authentication.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        if ($user) {
+            header("Location: ../views/dashboard.php");
+            exit();
+        } else {
+            header("Location: ../views/login.php?error=Incorrect Username or Password");
+            exit();
+        }
+    }
+
+    public function handleLoginRequest() {
```
Comment (bpocrafael) [inline html/login/controllers/authentication.php]: you could improve this just nit pick using the guard clause early return.

```

public function handleLoginRequest() {
    if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
        header("Location: ../views/login.php");
        exit();
    }

    if (!isset($_POST['username']) || !isset($_POST['password'])) {
         header("Location: ../views/login.php?error=" . urlencode($errorMessage));
    exit();
    }

    $username = $_POST['username'];
    $password = $_POST['password'];

    if (empty($username)) {
        $this->redirectToLoginPageWithMessage("Username is required");
    } 

    if (empty($password)) {
        $this->redirectToLoginPageWithMessage("Password is required");
    }

    $this->authenticate($username, $password);
}

```

File: html/login/controllers/authentication.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        if (empty($password)) {
+            header("Location: ../views/login.php?error=" . urlencode("Password is required"));
+            exit();
+        }
+
+        $this->authenticate($username, $password);
+    }   
+}
+
+$authController = new AuthenticationController();
```
Comment (hipe01mark) [inline html/login/controllers/authentication.php]: Codes that are executed outside the class. You can create __construct a function to initialize it instead of running it outside the class.

Reference:
https://www.w3schools.com/php/php_oop_constructor.asp#:~:text=PHP%20%2D%20The%20__construct%20Function,an%20object%20from%20a%20class.

**Review by hipe01mark — APPROVED** [review]
LGTM 🙇
